### PR TITLE
Support for overriding kubectl context during deployment

### PR DESCRIPTION
**Pull request by kaperys**
Adds an optional `context` property to skaffold.yaml. When provided, this value is used, instead of the current kubectl context.

References: #511

### CODE REVIEW

**Comment by kaperys** [comment]
I've amended this to include support for the `KUBE_CONTEXT` environment variable. The kubectl context can now be overridden using both the `KUBE_CONTEXT` environment variable and the context property in the skaffold.yaml schema. The order of precedence is environment, schema, kubectl context.

File: pkg/skaffold/schema/latest/config.go [hunk 1]
```diff
@@ -161,6 +181,8 @@ type DeployType struct {
 
 // KubectlDeploy contains the configuration needed for deploying with `kubectl apply`
 type KubectlDeploy struct {
+	Deployer `yaml:",inline"`
```
Comment (corneliusweig) [inline pkg/skaffold/schema/latest/config.go]: I find `Deployer` a bit misleading. Why not `Context`?

File: pkg/skaffold/schema/latest/config.go [hunk 2]
```diff
@@ -146,6 +148,24 @@ type TestCase struct {
 	StructureTests []string `yaml:"structureTests,omitempty"`
 }
 
+// Deployer contains properties shared between deployers
+type Deployer struct {
```
Comment (corneliusweig) [inline pkg/skaffold/schema/latest/config.go]: I find it a bit troubling that this type is defined in `latest/config.go`. This means that whenever a new api version is introduced, this logic (and the tests) need to be duplicated. What about moving this type into `package util` (`schema/util`)? The only drawback I see here is that the schema definition is no longer completely self-contained per api version.
@nkubala what do you think?

File: pkg/skaffold/schema/latest/config.go [hunk 3]
```diff
@@ -146,6 +148,24 @@ type TestCase struct {
...
+	KubeContext string `yaml:"context,omitempty"`
+}
+
+// ParseContext returns the kubectl context to use for deployment
+func (d Deployer) ParseContext(kubeContext string) string {
```
Comment (corneliusweig) [inline pkg/skaffold/schema/latest/config.go]: This doesn't really parse anything. Why not just `GetContext(..)` or even `Context(..)`?

Comment (r2d4) [reply pkg/skaffold/schema/latest/config.go]: ParseContext logic should be outside the config - most likely in https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/runner/runner.go#L64

File: pkg/skaffold/schema/latest/config.go [hunk 4]
```diff
@@ -177,11 +199,15 @@ type KubectlFlags struct {
 
 // HelmDeploy contains the configuration needed for deploying with helm
 type HelmDeploy struct {
+	Deployer `yaml:",inline"`
```
Comment (r2d4) [inline pkg/skaffold/schema/latest/config.go]: You don't need a new struct - you can just add `Context` to the DeployConfig since all deployers will be forced to respect it.

File: pkg/skaffold/schema/latest/config.go (hunk 2 above)
Comment (r2d4) [reply pkg/skaffold/schema/latest/config.go]: Not in schema/util but in pkg/runner, schema package should only be responsible the actual config schema operations

:YAML parsing, default values, profiles, config upgrade

**Comment by kaperys** [comment]
Thanks for your reviews, @r2d4, @corneliusweig and @nkubala! 

I've moved the context parsing logic into the runner package and removed the `Deployer` type; instead, merging the `KubeContext` property into `DeployConfig`. This way the schema definitions and properties remain in the schema package and the configuration logic belongs to the runner package.

File: docs/content/en/docs/concepts/_index.md [hunk 5]
```diff
@@ -44,6 +44,14 @@ For example, to treat any context as local by default:
 skaffold config set --global local-cluster true
 ```
 
+## Environment configuration
+
+Some environment or developer specific settings can be configured using environment variables. The options are
+
+| Variable Name | Description |
+| ------ | ----------- |
+| `KUBE_CONTEXT` | The kubectl context to use for deployment. |
```
Comment (corneliusweig) [inline docs/content/en/docs/concepts/_index.md]: I'm a bit worried about setting the kubecontext from the environment. There is a lengthy discussion about exactly that env variable around `kubectl`: https://github.com/kubernetes/kubernetes/issues/27308 . This would preclude a future command line flag `--context` which would be more in line with other kubernetes tools like `kubectl`.
@kaperys Does that solve a particular problem you have, or did you just add it for good measure?

File: pkg/skaffold/runner/runner.go [hunk 6]
```diff
@@ -170,13 +170,28 @@ func getTester(cfg []*latest.TestCase, opts *config.SkaffoldOptions) (test.Teste
...
+}
+
 func getDeployer(cfg *latest.DeployConfig, kubeContext string, namespace string, defaultRepo string) (deploy.Deployer, error) {
 	// TODO(dgageot): this should be the folder containing skaffold.yaml. Should also be moved elsewhere.
 	cwd, err := os.Getwd()
 	if err != nil {
 		return nil, errors.Wrap(err, "finding current directory")
 	}
 
+	kubeContext = getContext(kubeContext, cfg.KubeContext)
```
Comment (corneliusweig) [inline pkg/skaffold/runner/runner.go]: As already pointed out in the previous revision, `getContext` should be called one level up in 
`NewForConfig` (pkg/skaffold/runner/runner.go:64) because actual context also needs to be passed to `getBuilder` (pkg/skaffold/runner/runner.go:86).
Also, `GetLocalCluster` (cmd/skaffold/app/cmd/config/util.go:177) needs to know about the actual kubecontext in order to determine whether to treat the build as local or not. For that, you need to pass in `kubeContext` as an additional argument to `GetLocalCluster`.

File: pkg/skaffold/runner/runner_test.go [hunk 7]
```diff
@@ -321,3 +322,24 @@ func TestRun(t *testing.T) {
 		})
 	}
 }
+
+func TestGetContextEnvironment(t *testing.T) {
+	os.Setenv("KUBE_CONTEXT", "test-environment-context")
```
Comment (corneliusweig) [inline pkg/skaffold/runner/runner_test.go]: This may be an edge case, but this could alter the environment variables. To correctly restore the previous state, you need to query the current value, and restore that in the deferred cleanup.

File: pkg/skaffold/runner/runner_test.go [hunk 8]
```diff
@@ -321,3 +322,24 @@ func TestRun(t *testing.T) {
...
+	defer os.Unsetenv("KUBE_CONTEXT")
+
+	if c := getContext("test-kube-context", ""); c != "test-environment-context" {
+		t.Errorf("expected %q, have %q", "test-environment-context", c)
+	}
+}
+
+func TestParseContextDeployer(t *testing.T) {
+	if c := getContext("test-kube-context", "test-deployer-context"); c != "test-deployer-context" {
+		t.Errorf("expected %q, have %q", "test-deployer-context", c)
```
Comment (corneliusweig) [inline pkg/skaffold/runner/runner_test.go]: for assertions, you should use the standard skaffold utilities, e.g. `testutil.CheckDeepEqual`.

File: pkg/skaffold/runner/runner_test.go [hunk 9]
```diff
@@ -321,3 +322,24 @@ func TestRun(t *testing.T) {
...
+func TestGetContextEnvironment(t *testing.T) {
+	os.Setenv("KUBE_CONTEXT", "test-environment-context")
+	defer os.Unsetenv("KUBE_CONTEXT")
+
+	if c := getContext("test-kube-context", ""); c != "test-environment-context" {
+		t.Errorf("expected %q, have %q", "test-environment-context", c)
+	}
+}
+
+func TestParseContextDeployer(t *testing.T) {
```
Comment (corneliusweig) [inline pkg/skaffold/runner/runner_test.go]: nit: test name doesn't correspond to the method name

File: pkg/skaffold/runner/runner_test.go [hunk 10]
```diff
@@ -321,3 +322,24 @@ func TestRun(t *testing.T) {
...
+	if c := getContext("test-kube-context", "test-deployer-context"); c != "test-deployer-context" {
+		t.Errorf("expected %q, have %q", "test-deployer-context", c)
+	}
+}
+
+func TestParseContextKubeConfig(t *testing.T) {
```
Comment (corneliusweig) [inline pkg/skaffold/runner/runner_test.go]: nit: test name doesn't correspond to the method name

File: pkg/skaffold/runner/runner_test.go [hunk 11]
```diff
@@ -321,3 +322,24 @@ func TestRun(t *testing.T) {
 		})
 	}
 }
+
+func TestGetContextEnvironment(t *testing.T) {
+	os.Setenv("KUBE_CONTEXT", "test-environment-context")
+	defer os.Unsetenv("KUBE_CONTEXT")
+
+	if c := getContext("test-kube-context", ""); c != "test-environment-context" {
```
Comment (corneliusweig) [inline pkg/skaffold/runner/runner_test.go]: This test should make clear that the environment overrides the context from the schema. For that, the schema-context cannot be `""`.

File: docs/content/en/docs/concepts/_index.md (hunk 5 above)
Comment (kaperys) [reply docs/content/en/docs/concepts/_index.md]: Mostly for good measure - some earlier comments in #511 suggested this change. Supporting the `context` property in the skaffold.yml schema will solve my particular problem.

File: pkg/skaffold/runner/runner.go (hunk 6 above)
Comment (kaperys) [reply pkg/skaffold/runner/runner.go]: Apologies - I'd missed that. `GetLocalCluster` now accepts kubeContext.

File: pkg/skaffold/runner/runner.go [hunk 12]
```diff
@@ -170,6 +172,19 @@ func getTester(cfg []*latest.TestCase, opts *config.SkaffoldOptions) (test.Teste
 	}
 }
 
+// getContext returns the kubectl context to use for deployment
+func getContext(originalContext, schemaContext string) string {
```
Comment (corneliusweig) [inline pkg/skaffold/runner/runner.go]: This logic should be moved and merged into cmd/skaffold/app/cmd/config/util.go:36

File: pkg/skaffold/runner/runner.go [hunk 13]
```diff
@@ -170,6 +172,19 @@ func getTester(cfg []*latest.TestCase, opts *config.SkaffoldOptions) (test.Teste
...
+	if kc := os.Getenv("KUBE_CONTEXT"); kc != "" {
+		return kc
+	}
+
+	if kc := schemaContext; kc != "" {
```
Comment (corneliusweig) [inline pkg/skaffold/runner/runner.go]: This looks a bit weird. Why not
```
if schemaContext != nil {
  return schemaContext
}
```

File: cmd/skaffold/app/cmd/config/util.go [hunk 14]
```diff
@@ -94,12 +94,18 @@ func GetConfigForKubectx() (*ContextConfig, error) {
 	if global {
 		return cfg.Global, nil
 	}
+
+	localContext := kubecontext
```
Comment (corneliusweig) [inline cmd/skaffold/app/cmd/config/util.go]: TAG_TOO_LATE Too late, `resolveKubectlContext` could already have set the `global` flag (condition cmd/skaffold/app/cmd/config/util.go:45).

**Comment by kaperys** [comment]
@corneliusweig thank you for your review, and apologies for the confusion on my part. I've reset my branch and refactored out `overrideContext`, instead, adding a `GetKubeContext` function to cmd/skaffold/app/cmd/config/util.go as a "setup" method as you suggested. I've also removed the environment based override for now, until the discussion in [kubernetes/kubernetes#27308](https://github.com/kubernetes/kubernetes/issues/27308) is resolved. Please let me know if I have missed anything else.

**Comment by kaperys** [comment]
Thanks @corneliusweig, this much looks better. I had misunderstood your comment regarding setting the `kubecontext` global variable, which is why I added additional arguments to `GetDefaultRepo` and `NewForConfig`, etc - apologies. I'll fix up my PR now!

**Comment by corneliusweig** [comment]
@kaperys Please rebase this to resolve any conflicts.

**Comment by balopat** [comment]
Apologies @kaperys  - we'll need another rebase + regenerate schemas. I added the v1beta7 schema where this change should land. Thank you!

**Comment by kaperys** [comment]
@balopat I've rebased and pushed. Thanks!

File: cmd/skaffold/app/cmd/config/util.go [hunk 15]
```diff
@@ -33,22 +33,40 @@ import (
...
 
+	if schemaContext != "" {
+		kubecontext = schemaContext
+		return
+	}
+
 	context, err := context.CurrentContext()
 	if err != nil {
-		logrus.Warn(errors.Wrap(err, "retrieving current kubectl context"))
+		logrus.Warn(errors.Wrapf(err, "retrieving current kubectl context"))
```
Comment (balopat) [inline cmd/skaffold/app/cmd/config/util.go]: why Wrapf?

File: cmd/skaffold/app/cmd/config/util.go [hunk 16]
```diff
@@ -33,22 +33,40 @@ import (
 const defaultConfigDir = ".skaffold"
 const defaultConfigFile = "config"
 
-func resolveKubectlContext() {
+// ResolveKubectlContext sets up the actual kube-context to be used. The order
+// of precedence is:
+// 1) command line
+// 2) skaffold.yml
```
Comment (balopat) [inline cmd/skaffold/app/cmd/config/util.go]:
```suggestion
// 2) skaffold.yaml
```

File: cmd/skaffold/app/cmd/config/util.go [hunk 17]
```diff
@@ -33,22 +33,40 @@ import (
 const defaultConfigDir = ".skaffold"
 const defaultConfigFile = "config"
 
-func resolveKubectlContext() {
+// ResolveKubectlContext sets up the actual kube-context to be used. The order
+// of precedence is:
+// 1) command line
```
Comment (balopat) [inline cmd/skaffold/app/cmd/config/util.go]: this is a misunderstanding - there is _no command line flag yet_ for kubectl context. See below, the kubecontext you are using is a global var used to parse the `skaffold config` command's kubeContext flag, so that the `skaffold config` can store config values scoped to a given kubectl context.

File: cmd/skaffold/app/cmd/config/util.go [hunk 18]
```diff
@@ -33,22 +33,40 @@ import (
...
+// 2) skaffold.yml
+// 3) current context from kube-config
+func ResolveKubectlContext(schemaContext string) {
 	if kubecontext != "" {
 		return
 	}
 
+	if schemaContext != "" {
```
Comment (balopat) [inline cmd/skaffold/app/cmd/config/util.go]: 1. this whole logic should not be under the `config` cmd. 
2. nit: `schemaContext` I would rename to `deployConfigContext`

File: cmd/skaffold/app/cmd/config/util.go (hunk 17 above)
Comment (kaperys) [reply cmd/skaffold/app/cmd/config/util.go]: To clarify, `config.kubecontext` shouldn't be considered as part of this PR, then? Ie, when using commands like `dev` and `run` the only contexts that should be considered are the schema.yaml context and the current kubectl context?

Comment (balopat) [reply cmd/skaffold/app/cmd/config/util.go]: right - there are two flows here that are affected ... and as I am digging deeper, this is not that trivial actually. 

1.) "runner" (`skaffold dev/deploy/run`): 

a.) we need some metadata - namely `config.GetLocalCluster` and `GetDefaultRepo` to influence right behavior for artifact cache and image replacement. These both rely on the "right kubectl context". 
b.) We need the "right context" for profile activation. 
c.) the deployer (helm/kubectl) needs to be created with the "right context".
d.) bazel and local docker has some kube context dependent logic. 

Ideally we need a "kubectx.CurrentContext()" to return the "right context" that is not the current kubectx on the machine, but whatever we calculated from config. 

The logic to determine the "right kubectl context" should be based on precedence of `DeployConfig.context` and `kubectl.CurrentContext()` - however there is a bit of a circular dependency now that is problematic because of the profile activations (@dgageot???) - we can theoretically create activate a profile based on the current context that will have a Deploy config pointing to a different context. The way to resolve this would be to implement validation against the skaffold.yaml: the kubecontext under deployConfig should never be different from the profile it is defined under. 

Thus the logic to determine the would be the following: 
- get current kubectl context (from machine) 
- if there exists a profile that is activated by this context - make this the current context for everybody - do validate the deploy configs against mismatching context definition 
- if there is no profile that is activated - check the DeployConfig.context - if it's defined, make that the current context for all the above (a, c, d) clients. 

One other thought I have is that if we add multiple kubecontext support for deployers - we might have to rethink what this global context means - but I think we can do that based on this later - so we can ignore that use case for now. 

**problems with current solution**: 

- does not need the "kubecontext" flag from `skaffold config` - as it is never called. 
- the fact that you are storing the result of the resolution in a flag's value that is disconnected also should be reconsidered as everyone else is calling `context.CurrentContext()` which will result in inconsistencies - I realize that config package is 

**recommendation**: 
- implement the above logic for determining the kubectx in a function named ResolveAndValidateKubeCtx(config) `pkg/skaffold/kubernetes/context/context.go` 
- this _can_ set a state variable in `pkg/skaffold/kubernetes/context`
- `ResolveAndValidateKubeCtx` should be called right after the config has been parsed - maybe even part of the validation there? 
- `context.CurrentContext()`  should always just return the value of the state variable if it's set, if not, it should return the global/machine kubectl context as it does now

2.) `skaffold config` subcommands: have to resolve the right "kubectl context" based on the precedence of `--kube-context` flag (or KUBE_CONTEXT env var -- see `cmd.setFlagsFromEnvVariables`) and `kubectl.CurrentContext()` to retrieve/set/unset the right configuration in the global config. 

**problems with current solution**: 

- does not need the `DeployConfig.context` as `skaffold config` doesn't care about the skaffold.yaml file itself. 

**recommendation**: 
- leave it as is for now, and the rest is going to be taken care of as `context.CurrentContext()`  should always just return the value of the state variable if it's set, if not, it should return the global/machine kubectl context as it does now

**Comment by temiwottku** [comment]
Hi!
Any news on this PR? I hope it wasn't abandoned, it's quite the useful feature!

**Comment by 1N50MN14** [comment]
Any progress on this PR @kaperys @balopat?

I was just going through the docs looking if there was a way to pin a kube context to a specific configuration (basically `deploy.kubectl.context`) to avoid accidentally deploying the wrong cluster, as a user this was an instinctive thing for me to look for.

<3 skaffold

**Comment by nkubala** [comment]
we're definitely in favor of this feature being developed, but in its current state I think this PR needs to be closed due to inactivity. @balopat's comment [here](https://github.com/GoogleContainerTools/skaffold/pull/1540#discussion_r265722884) provides a good starting point for implementing this, but I think this PR is at risk for falling too far behind current master, so I'm going to close this for now. if anybody wants to take another stab at implementing this PRs are of course welcome!
